src/http/matcher/mod.rs

```diff
@@ -65,9 +68,30 @@ pub enum HttpMatcherKind {
     ///
     /// [`SocketAddr`]: std::net::SocketAddr
     Socket(SocketMatcher),
+    /// A custom matcher that implements [`crate::service::Matcher`].
+    // Custom(Box<dyn crate::service::Matcher<State, Request<Body>>>), TODO: Does not satisfy Clone, unless inner field satisfies Clone.
```

Comment:
We would need `dyn crate::service::Matcher<State, Request<Body>>` to impl `Clone` if using `Box`.

Reply:
Using `Arc` for now.

Reply:
I prefer`Box`, so far I've been doing that in `rama`, so for consistency would be better and also signals better that it's a trait object.

If a matcher has to be clone-able they can wrap the entire `Matcher` with an `Arc`.

Reply:
Nevermind @dmackdev , you make me happy. The `Arc` usage here is actually genious. Keep it.

However in order to resolve your compile errors in `rama-fp` you'll have to remove the `#[derive(Debug, Clone)]` statements for the `HttpMatcherKind` and `HttpMatcher`.

You can then just implement them manually, e.g. for Clone that means:

```rust
impl<State, Body> Clone for HttpMatcher<State, Body> {
    fn clone(&self) -> Self {
        Self {
            kind: self.kind.clone(),
            negate: self.negate,
        }
    }
}

impl<State, Body> Clone for HttpMatcherKind<State, Body> {
    fn clone(&self) -> Self {
        match self {
            Self::All(v) => Self::All(v.clone()),
            Self::Method(v) => Self::Method(v.clone()),
            Self::Path(v) => Self::Path(v.clone()),
            Self::Domain(v) => Self::Domain(v.clone()),
            Self::Version(v) => Self::Version(v.clone()),
            Self::Any(v) => Self::Any(v.clone()),
            Self::Uri(v) => Self::Uri(v.clone()),
            Self::Header(v) => Self::Header(v.clone()),
            Self::Socket(v) => Self::Socket(v.clone()),
            Self::Custom(v) => Self::Custom(v.clone()),
        }
    }
}
```

And Debug you can implement just the same way. But there you might need to add a `where` clause to require Debug .Not sure.

Reply:
Done.

You are correct that `where State: fmt::Debug, Body: fmt::Debug` was needed to implement `Debug`.